snapcraft/commands/lifecycle.py

```diff
@@ -17,16 +17,107 @@
+    overview = textwrap.dedent(
+        """
+        Process parts and create a snap file containing the project payload
+        with the provided metadata. If a directory is specified, pack its
+        contents instead.
+        """
+    )
+
+    @override
+    def _fill_parser(self, parser: argparse.ArgumentParser) -> None:
+        """Add arguments specific to the pack command."""
+        super()._fill_parser(parser)
+
+        parser.add_argument(
+            "directory",
+            metavar="directory",
+            type=str,
+            nargs="?",
+            default=None,
+            help="Directory to pack",
+        )
+        parser.add_argument(
+            "--output",
+            "-o",
+            type=pathlib.Path,
+            default=pathlib.Path(),
+            help="Output directory for created packages",
+        )
+
+    @override
+    def _run(
+        self,
+        parsed_args: argparse.Namespace,
+        step_name: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Pack a directory or run the lifecycle and pack all artifacts."""
+        if parsed_args.directory:
+            emit.progress("Packing...")
+            snap_filename = snapcraft.pack.pack_snap(
+                parsed_args.directory, output=str(parsed_args.output)
+            )
+            emit.message(f"Packed {snap_filename}")
+        else:
+            super()._run(parsed_args, step_name="prime")
+            self._services.package.update_project()
+            self._services.package.write_metadata(self._services.lifecycle.prime_dir)
+
+            emit.progress("Packing...")
+            packages = self._services.package.pack(
+                self._services.lifecycle.prime_dir, parsed_args.output
+            )
+
+            if not packages:
+                emit.progress("No packages created.", permanent=True)
+            elif len(packages) == 1:
+                emit.progress(f"Packed {packages[0].name}", permanent=True)
+            else:
+                package_names = ", ".join(pkg.name for pkg in packages)
+                emit.progress(f"Packed: {package_names}", permanent=True)
+
+    @override
+    def needs_project(self, parsed_args: argparse.Namespace) -> bool:
+        """Project is not required to pack a directory."""
+        if parsed_args.directory:
+            emit.debug("Not loading project because a directory was provided.")
+            return False
+
+        emit.debug("Loading project because a directory was not provided.")
+        return True
+
+    @override
+    def run_managed(self, parsed_args: argparse.Namespace) -> bool:
+        """Return whether the command should run in managed mode or not.
+
+        Packing a directory always runs locally.
+        """
+        if parsed_args.directory:
+            emit.debug("Not running managed mode because a directory was provided.")
+            return False
+
+        return super().run_managed(parsed_args)
```

Comment:
I think something like this should work:

```suggestion
class PackCommand(craft_application.commands.PackCommand):
    """Snapcraft pack command."""

    overview = textwrap.dedent(
        """
        Process parts and create a snap file containing the project payload
        with the provided metadata. If a directory is specified, pack its
        contents instead.
        """
    )

    @override
    def _fill_parser(self, parser: argparse.ArgumentParser) -> None:
        """Add arguments specific to the pack command."""
        # Don't add parts arguments.
        super(craft_application.commands.LifecycleCommand)._fill_parser(parser)
        
       parser.add_argument(
            "directory",
            metavar="directory",
            type=str,
            nargs="?",
            default=None,
            help="Directory to pack",
        )
        parser.add_argument(
            "--output",
            "-o",
            type=pathlib.Path,
            default=pathlib.Path(),
            help="Output directory for created packages",
        )

    @override
    def _run(
        self,
        parsed_args: argparse.Namespace,
        step_name: str | None = None,
        **kwargs: Any,
    ) -> None:
        """Pack a directory or run the lifecycle and pack all artifacts."""
        if parsed_args.directory:
            emit.progress("Packing...")
            snap_filename = snapcraft.pack.pack_snap(
                parsed_args.directory, output=str(parsed_args.output)
            )
            emit.message(f"Packed {snap_filename}")
        else:
            parsed_args.parts = None
	        super()._run(parsed_args)

    @override
    def needs_project(self, parsed_args: argparse.Namespace) -> bool:
        """Project is not required to pack a directory."""
        if parsed_args.directory:
            emit.debug("Not loading project because a directory was provided.")
            return False

        emit.debug("Loading project because a directory was not provided.")
        return True

    @override
    def run_managed(self, parsed_args: argparse.Namespace) -> bool:
        """Return whether the command should run in managed mode or not.

        Packing a directory always runs locally.
        """
        if parsed_args.directory:
            emit.debug("Not running managed mode because a directory was provided.")
            return False

        return super().run_managed(parsed_args)
```

Reply:
I like your idea and I played with it for a bit but am struggling with the `super(craft_application.commands.LifecycleCommand)._fill_parser(parser)` call.

Here's what I will do - I'm going to do another PR in craft-application where I change the parent of the `PackCommand` and then remove this duplicated code in snapcraft.  I'll do it on the `main` branch and it will be part of snapcraft 8.3.